tests/api/utils/test_pagination.py

```diff
@@ -0,0 +1,363 @@
+    response, pagination_info = paginator.paginate_request(
+        db, paginated_method, auth_info, pagination_info["token"]
+    )
+    _assert_paginated_response(
+        response, pagination_info, 2, page_size, ["item3", "item4"]
+    )
+
+    logger.info("Checking db cache record")
+    cache_record = server.api.crud.PaginationCache().get_pagination_cache_record(
+        db, pagination_info["token"]
+    )
+    _assert_cache_record(
+        cache_record, auth_info.user_id, paginated_method, 2, page_size
+    )
+
+    logger.info("Saving token for next assert")
+    token = pagination_info["token"]
+
+    logger.info(
+        "Requesting third page, which is the end of the items and should return empty response"
+    )
+    response, pagination_info = paginator.paginate_request(
+        db, paginated_method, auth_info, pagination_info["token"]
+    )
+    assert len(response) == 0
+    assert not pagination_info
+
+    logger.info("Checking db cache record was removed")
+    cache_record = server.api.crud.PaginationCache().get_pagination_cache_record(
+        db, token
+    )
+    assert cache_record is None
+
+
+def test_paginate_other_users_token(mock_paginated_method, db: sqlalchemy.orm.Session):
+    """
+    Test pagination with a token that was created by a different user.
+    Request paginated method with page and page size, and verify that the correct items are returned.
+    Check the db for the pagination cache record.
+    Request the next page with the token, and with different user, and verify that a AccessDeniedError is raised.
+    """
+    auth_info_1 = mlrun.common.schemas.AuthInfo(user_id="user1")
+    auth_info_2 = mlrun.common.schemas.AuthInfo(user_id="user2")
+    page_size = 3
+    method_kwargs = {"total_amount": 5}
+
+    paginator = server.api.utils.pagination.Paginator()
+
+    logger.info("Requesting first page with user1")
+    response, pagination_info = paginator.paginate_request(
+        db, paginated_method, auth_info_1, None, 1, page_size, **method_kwargs
+    )
+    _assert_paginated_response(
+        response, pagination_info, 1, page_size, ["item0", "item1", "item2"]
+    )
+
+    logger.info("Checking db cache record")
+    cache_record = server.api.crud.PaginationCache().get_pagination_cache_record(
+        db, pagination_info["token"]
+    )
+    _assert_cache_record(
+        cache_record, auth_info_1.user_id, paginated_method, 1, page_size
+    )
+
+    logger.info("Requesting second page with user2, should raise AccessDeniedError")
+    with pytest.raises(mlrun.errors.MLRunAccessDeniedError):
+        paginator.paginate_request(
+            db, paginated_method, auth_info_2, pagination_info["token"]
+        )
+
+    logger.info(
+        "Requesting second page without auth info, should raise AccessDeniedError"
+    )
+    with pytest.raises(mlrun.errors.MLRunAccessDeniedError):
+        paginator.paginate_request(db, paginated_method, None, pagination_info["token"])
```

Comment:
What about cleaning up the pagination cache? Aren't you concerned about pagination cache leaks across tests?

Reply:
Not really as we delete the db when the test goes down. But can add it to the cleanup